plugins/lang-plan-a/src/test/java/org/elasticsearch/plan/a/WhenThingsGoWrongTests.java

```diff
@@ -49,4 +49,62 @@ public void testBogusParameter() {
             assertTrue(expected.getMessage().contains("Unrecognized compile-time parameter"));
         }
     }
+
+    public void testInfiniteLoops() {
+        try {
+            exec("boolean x = true; while (x) {}");
+            fail("should have hit PlanAError");
+        } catch (PlanAError expected) {
+            assertTrue(expected.getMessage().contains(
+                "The maximum number of statements that can be executed in a loop has been reached."));
+        }
+
+        try {
+            exec("while (true) {int y = 5}");
+            fail("should have hit PlanAError");
+        } catch (PlanAError expected) {
+            assertTrue(expected.getMessage().contains(
+                "The maximum number of statements that can be executed in a loop has been reached."));
+        }
+
+        try {
+            exec("while (true) { boolean x = true; while (x) {} }");
+            fail("should have hit PlanAError");
+        } catch (PlanAError expected) {
+            assertTrue(expected.getMessage().contains(
+                "The maximum number of statements that can be executed in a loop has been reached."));
+        }
+
+        try {
+            exec("while (true) { boolean x = false; while (x) {} }");
+            fail("should have hit PlanAError");
+        } catch (PlanAError expected) {
+            assertTrue(expected.getMessage().contains(
+                "The maximum number of statements that can be executed in a loop has been reached."));
+        }
+
+        try {
+            exec("boolean x = true; for (;x;) {}");
+            fail("should have hit PlanAError");
+        } catch (PlanAError expected) {
+            assertTrue(expected.getMessage().contains(
+                "The maximum number of statements that can be executed in a loop has been reached."));
+        }
+
+        try {
+            exec("for (;;) {int x = 5}");
+            fail("should have hit PlanAError");
+        } catch (PlanAError expected) {
+            assertTrue(expected.getMessage().contains(
+                "The maximum number of statements that can be executed in a loop has been reached."));
+        }
+
+        try {
+            exec("def x = true; do {int y = 5;} while (x)");
+            fail("should have hit PlanAError");
+        } catch (PlanAError expected) {
+            assertTrue(expected.getMessage().contains(
+                "The maximum number of statements that can be executed in a loop has been reached."));
+        }
+    }
```

Comment:
Maybe we should test around the boundaries of the loop limit (rather than just infinite ones, finite-but-large ones) to ensure it works exactly as we expect?